File: Day1/src/domain/RPSHand.java
package domain;

public enum RPSHand {
    ROCK(1),
    PAPER(2),
    SCISSORS(3);

    private final int points;

    RPSHand(int points) {
        this.points = points;
    }

    public int getPoints() {
        return points;
    }

    public static RPSHand getMatching(String variable) {
        switch(variable) {
            case "A":
            case "X":
                return ROCK;
            case "B":
            case "Y":
                return PAPER;
            case "C":
            case "Z":
                return SCISSORS;
            default:
                throw new IllegalArgumentException();
        }
    }

    public int getPointsAgainst(RPSHand other) {
        if (this == other) return 3;
        switch (this) {
            case ROCK:
                return other == PAPER ? 0 : 6;
            case PAPER:
                return other == SCISSORS ? 0 : 6;
            case SCISSORS:
                return other == ROCK ? 0 : 6;
            default:
                throw new IllegalArgumentException();
        }
    }

    private RPSHand getWinner() {
        switch(this) {
            case ROCK: return PAPER;
            case PAPER: return SCISSORS;
            case SCISSORS: return ROCK;
            default:
                throw new IllegalArgumentException();
        }
    }

    private RPSHand getLoser() {
        switch(this) {
            case ROCK: return SCISSORS;
            case PAPER: return ROCK;
            case SCISSORS: return PAPER;
            default:
                throw new IllegalArgumentException();
        }
    }

    public RPSHand getHandFor(String result) {
        switch(result) {
            case "X":
                return getLoser();
            case "Y":
                return this;
            case "Z":
                return getWinner();
            default:
                throw new IllegalArgumentException();
        }
    }
}

File: Day1/src/domain/RPSHand.kt
package domain

enum class RPSHand(val points: Int) {
    ROCK(1), PAPER(2), SCISSORS(3);

    fun getPointsAgainst(other: RPSHand): Int {
        return if (this == other) 3 else when (this) {
            ROCK -> if (other == PAPER) 0 else 6
            PAPER -> if (other == SCISSORS) 0 else 6
            SCISSORS -> if (other == ROCK) 0 else 6
        }
    }

    fun getHandFor(result: String?): RPSHand {
        return when (result) {
            "X" -> loser(this)
            "Y" -> this
            "Z" -> winner(this)
            else -> throw IllegalArgumentException()
        }
    }

    companion object {
        fun getMatching(variable: String?): RPSHand {
            return when (variable) {
                "A", "X" -> ROCK
                "B", "Y" -> PAPER
                "C", "Z" -> SCISSORS
                else -> throw IllegalArgumentException()
            }
        }

        private fun winner(hand: RPSHand): RPSHand = when (hand) {
            ROCK -> PAPER
            PAPER -> SCISSORS
            SCISSORS -> ROCK
        }

        private fun loser(hand: RPSHand): RPSHand = when (hand) {
            ROCK -> SCISSORS
            PAPER -> ROCK
            SCISSORS -> PAPER
        }
    }
}